tools/serverpod_cli/lib/src/database/migration.dart

```diff
@@ -11,14 +12,26 @@ DatabaseMigration generateDatabaseMigration({
 
   // Find deleted tables
   var deleteTables = <String>[];
-  var sourceTables = databaseSource.tables.where((table) => table.isManaged);
+  var sourceTables =
+      databaseSource.tables.where((table) => table.isManaged).toList();
   var targetTables = databaseTarget.tables.where((table) => table.isManaged);
 
-  for (var srcTable in sourceTables) {
-    if (!databaseTarget.containsTableNamed(srcTable.name)) {
-      deleteTables.add(srcTable.name);
+  for (var i = 0; i < sourceTables.length; i++) {
+    var srcTable = sourceTables[i];
+
+    if (!databaseTarget.containsTableNamed(srcTable.name) ||
+        srcTable.foreignKeys.any(
+            (foreignKey) => deleteTables.contains(foreignKey.referenceTable))) {
+      if (!deleteTables.contains(srcTable.name)) {
+        deleteTables.add(srcTable.name);
+
+        // Start the loop over, as an earlier table might point to the newly deleted
+        // table, and thus also needs to get dropped now.
+        i = -1; // loop will add +1 again, making this 0
+      }
     }
   }
+
   for (var tableName in deleteTables.reversed) {
```

Comment:
Would've loved to understand why this is being forked on in reverse (as the `sourceTables` and thus `deleteTables` does not have a special order as far as I can see).